src/protocol/frame/init_aware_buf.rs

```diff
@@ -0,0 +1,210 @@
+use bytes::{Buf, BytesMut};
+use std::ops::{Deref, DerefMut};
+
+/// Buffer that provides fast & safe [`Self::resize`] & [`Self::truncate`] usage.
+///
+/// It is aware of the initialization state of its spare capacity avoiding the
+/// need to zero uninitialized bytes on resizing more than once for safe usage.
+///
+/// This optimisation is useful for [`std::io::Read`] to safely provide spare
+/// capacity as an initialized slice.
+///
+/// Related, may be obsoleted by: <https://github.com/rust-lang/rust/issues/78485>
+#[derive(Debug, Default)]
+pub struct InitAwareBuf {
+    /// Backing buf length is used as capacity. This ensure this extra region
+    /// is always initialized (initially with zero, but otherwise the last previously
+    /// set value).
+    bytes: BytesMut,
+    /// Length of bytes in use (always <= bytes.len).
+    len: usize,
+}
+
+impl InitAwareBuf {
+    #[inline]
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self { bytes: BytesMut::zeroed(capacity), len: 0 }
+    }
+
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.len
+    }
+
+    /// Capacity that may be resized to cheaply.
+    #[inline]
+    pub fn capacity(&self) -> usize {
+        self.bytes.len()
+    }
+
+    #[inline]
+    pub fn split_to(&mut self, at: usize) -> BytesMut {
+        assert!(at <= self.len, "split_to out of bounds: {at} <= {}", self.len);
+        let split = self.bytes.split_to(at);
+        self.len -= at;
+        split
+    }
+
+    /// Reserve capacity for `min_additional` more bytes than the current [`Self::len`].
+    ///
+    /// `max_additional` sets the maximum number of additional bytes zeroed as extra
+    /// capacity if available after reserving in the underlying buffer. Has no effect
+    /// if `max_additional <= additional`.
+    #[inline]
+    pub fn reserve(&mut self, additional: usize, max_additional: usize) {
+        let min_len = self.len + additional;
+        let cap = self.capacity();
+        if min_len > cap {
+            self.bytes.reserve(min_len - cap);
+            let new_cap = self.bytes.capacity().min(self.len + max_additional.max(additional));
+            self.bytes.resize(new_cap, 0);
+        }
+    }
+
+    /// Resizes the buffer to `new_len`.
+    ///
+    /// If greater the new bytes will be either initialized to zero or as
+    /// they were last set to.
+    #[inline]
+    pub fn resize(&mut self, new_len: usize) {
+        if new_len > self.capacity() {
+            self.bytes.resize(new_len, 0);
+        }
```

Comment:
Could just `assert!...` here as we don't expect this branch to execute but this reflects BytesMut functionality better...